plugins/out_s3/s3.c

```diff
@@ -1343,6 +1348,169 @@ static void cb_s3_upload(struct flb_config *config, void *data)
+        flb_errno();
+        return NULL;
+    }
+
+    /* Create temporary msgpack buffer */
+    msgpack_sbuffer_init(&tmp_sbuf);
+
+    msgpack_unpacked_init(&result);
+    while (!alloc_error && 
+           msgpack_unpack_next(&result, data, bytes, &off) == MSGPACK_UNPACK_SUCCESS) {
+        /* Each array must have two entries: time and record */
+        root = result.data;
+        if (root.type != MSGPACK_OBJECT_ARRAY) {
+            continue;
+        }
+        if (root.via.array.size != 2) {
+            continue;
+        }
+
+        /* Get the record/map */
+        map = root.via.array.ptr[1];
+        if (map.type != MSGPACK_OBJECT_MAP) {
+            continue;
+        }
+        map_size = map.via.map.size;
+
+        /* Extract log_key from record and append to output buffer */
+        for (i = 0; i < map_size; i++) {
+            key = map.via.map.ptr[i].key;
+            val = map.via.map.ptr[i].val;
+
+            if (key.type == MSGPACK_OBJECT_BIN) {
+                key_str  = (char *) key.via.bin.ptr;
+                key_str_size = key.via.bin.size;
+                check = FLB_TRUE;
+            }
+            if (key.type == MSGPACK_OBJECT_STR) {
+                key_str  = (char *) key.via.str.ptr;
+                key_str_size = key.via.str.size;
+                check = FLB_TRUE;
+            }
+
+            if (check == FLB_TRUE) {
+                if (strncmp(ctx->log_key, key_str, key_str_size) == 0) {
+                    found = FLB_TRUE;
+
+                    /* 
+                     * Copy contents of value into buffer. Necessary to strncpy
+                     * strings because flb_msgpack_to_json does not handle nested
+                     * JSON gracefully and double escapes them.
+                     */
+                    if (val.type == MSGPACK_OBJECT_BIN) {
+                        strncpy(val_buf, val.via.bin.ptr, val.via.bin.size);
+                        val_buf[val.via.bin.size] = '\0';
+                    }
+                    else if (val.type == MSGPACK_OBJECT_STR) {
+                        strncpy(val_buf, val.via.str.ptr, val.via.str.size);
+                        val_buf[val.via.str.size] = '\0';
+                    }
+                    else {
+                        ret = flb_msgpack_to_json(val_buf, msgpack_size, &val);
+                        if (ret < 0) {
+                            break;
+                        }
+                    }
+
+                    /* Create or concatenate to out_buf for full record chunk */
+                    if (out_buf == NULL) {
+                        out_buf = flb_sds_create(val_buf);
+                        if (out_buf == NULL) {
+                            flb_plg_error(ctx->ins, "Error creating buffer to "
+                                          "store log_key contents.");
+                            flb_errno();
+                            alloc_error = 1;
+                            break;
```

Comment:
I see alloc_error as a condition in the loop, but you also use a `break` (which seems better)...

Reply:
Because this part is in a nested loop, I need to double break, hence `alloc_error` to break out of the outer while loop. I could use a goto, but I felt a conditional variable was better practice here!

Reply:
I see. I think that works. I've also seen the use of `goto` with a well-named label like:

```
if something bad happens:
     goto break_outer_loop
```